README.md
<a href="http://promisesaplus.com/">
  <img src="https://promises-aplus.github.io/promises-spec/assets/logo-small.png"
    align="right" valign="top" alt="Promises/A+ logo">
</a>

# supertest-as-promised

<a href="https://travis-ci.org/WhoopInc/supertest-as-promised">
  <img src="https://travis-ci.org/WhoopInc/supertest-as-promised.svg?branch=master"
    align="right" valign="top" alt="Build Status">
</a>

**This project is deprecated!** As of v2.0.0, [SuperTest] has native
support for promises. **Don't use SuperTest as Promised in new projects**;
use SuperTest directly!

If you're currently using SuperTest as Promised, here's how to switch to
native SuperTest:

   * SuperTest returns only ES6 promises. SuperTest as Promised, by
     contrast, returns [Bluebird] promises by default, or promise
     objects from a library of your choosing using [BYOP
     support](#byop-bring-your-own-promise). Every promise library worth
     its salt supports ES6 promise interop, though, so it should be a
     small matter to cast ES6 promises to your custom promise library's
     promises. Here's a simple SuperTest as Promised snippet that relies
     on the `.delay` method of Bluebird promises

    ```js
    var request = require("supertest-as-promised");

    request(app)
      .get("/kittens")
      .expect(201)
      .toPromise()
      .delay(10)
      .then(function (res) { /* ... */ })
    ```

    that can be trivially converted to use SuperTest 2.0 like so:

    ```js
    var request = require("supertest");
    var BluebirdPromise = require("bluebird");

    BluebirdPromise.resolve(
      request(app)
        .get("/kittens")
        .expect(201))
      .delay(10)
      .then(function (res) { /* ... */ })
    ```

   * SuperTest does not ship a `.toPromise` method. You'll need to
     remove calls to this method. You were probably using `.toPromise`
     because you wanted a proper Promise object.
     `YourPromiseLibrary.resolve(...)` will probably do the trick; see
     the example above.

SuperTest as Promised will continue to receive bugfixes, but no new
features. Cheers, folks! It's been a good two years.

## Overview

SuperTest as Promised supercharges [SuperTest] with a `then` method.

Instead of layering callbacks on callbacks in your tests:

```js
request(app)
  .get("/user")
  .expect(200, function (err, res) {
    if (err) return done(err);

    var userId = res.body.id;
    request(app)
      .post("/kittens")
      .send({ userId: userId, ... })
      .expect(201, function (err, res) {
        if (err) return done(err);

        // ...
      });
  });
```

chain your requests like you were promised:

```js
return request(app)
  .get("/user")
  .expect(200)
  .then(function (res) {
    return request(app)
      .post("/kittens")
      .send({ userId: res})
      .expect(201);
  })
  .then(function (res) {
    // ...
  });
```

## Usage

SuperTest as Promised operates just like normal [SuperTest], except that the
object returned by `.get`, `.post`, etc. is a proper
thenable:

```js
var express = require("express")
  , request = require("supertest-as-promised");

var app = express();

request(app)
  .get("/kittens")
  .expect(200)
  .then(function (res) {
    // ...
  });
```

If you use a promise-friendly test runner, you can just
return your `request` chain from the test case rather than messing with a
callback:

```js
describe("GET /kittens", function () {
  it("should work", function () {
    return request(app).get("/kittens").expect(200);
  });
});
```

### Upstream documentation

SuperTest as Promised is built on top of [SuperTest], which is in turn
built on top of [SuperAgent]. You'll want to be familiar with both of
those packages' APIs.

Here's the quick rundown if you're unfamiliar. SuperAgent is a generic
HTTP client that provides methods for fluently constructing HTTP requests
(e.g., `.get`, `.send`, `.query`, `.auth`). See the
[SuperAgent docs][SuperAgent] for a full list of methods. SuperTest allows
you to tack on assertions about e.g. response status (`.expect(200)`) or
content (`.expect(/^kitties are \w+$/)`) to a SuperAgent request. See the
[`.expect` function API][expect-api] in the SuperTest documentation for details.

### Agents

If you use a SuperTest agent to persist cookies, those are thenable too:

```js
var agent = require("supertest-as-promised").agent(app);

agent
  .get("/ugly-kitteh")
  .expect(404)
  .then(function () {
    // ...
  })
```


### Promisey goodness

To start, only the `then` and `catch` methods are exposed. But once you've
called `.then` or `.catch` once, you've got a proper [Bluebird] promise that
supports the whole gamut of promisey goodness:

```js
request(app)
  .get("/kittens")
  .expect(201)
  .then(function (res) { /* ... */ })
  // I'm a real promise now!
  .catch(function (err) { /* ... */ })
```

See the [Bluebird API][bluebird-api] for everything that's available.

You may find it cleaner to cast directly to a promise using the `toPromise`
method:

```js
request(app)
  .get("/kittens")
  .expect(201)
  .toPromise()
  // I'm a real promise now!
  .delay(10)
  .then(function (res) { /* ... */ })
```

### BYOP: Bring your own `Promise`

You can supply own promise library so that the promises returned have your
convenience methods of choice.

Simply call the SuperTest as Promised module with a ES6-compliant `Promise`
constructor, and you'll get back a new module configured to return your custom
promises. To swap in [when.js], for example:

```js
var when = require("when")
  , request;

request = require("supertest-as-promised")(when.Promise);
request(app)
  .get("/when.js")
  .then(function (res) { /* ... */ })
  // I'm a when.js promise! (instanceof when.Promise == true)
  .frobulate()

request = require("supertest-as-promised");
request(app)
  .get("/bluebird.js")
  .then(function (res) { /* .. */ })
  // I'm back to the default Bluebird promise!
```

### Debugging

Suppose your snazzy test

```js
it("should get kittens", function () {
  return request(app).get("/kittens").expect(200)
    .then(function (res) {
      return expect(res.body).to.equal("kittens r cute!");
    });
});
```

suddenly starts failing with a "400 Bad Request" error. It'd sure be
handy if you could print out the response body to see if the server
mentioned what, in particular, was bad about your test. Chain on a
`.catch` and inspect `err.response`:

```js
it("should get kittens", function () {
  return request(app).get("/kittens").expect(200)
    .then(function (res) {
      return expect(res.text).to.equal("kittens r cute!");
    })
    .catch(function (err) {
      console.log(err.response.text); // "You have viewed too many kittens today."
    });
});
```


## Installation

### Node

```bash
$ npm install supertest supertest-as-promised
```

SuperTest as Promised lists [`supertest`][SuperTest] as a
[peer dependency][peer-dependency], so it'll wrap whatever version of SuperTest
you've asked for in your own `package.json`.

In earlier versions of NPM, failing to list SuperTest as a dependency would get
you the latest version. In NPM 3, failing to list SuperTest as a dependency will
generate a warning and SuperTest will not be installed.

Do note that SuperTest as Promised is a well-behaved citizen and doesn't
monkey-patch SuperTest directly:

```js
// I return thenables!
var request = require("supertest-as-promised");

// I'm lame and force you to use callbacks
var request = require("supertest");
```


## Versions

We follow [semver]: the major version number will be upgraded with any breaking
change. Breaking changes in each major version are listed below. Consult the
[changelog] for a list of meaningful new features in each version; consult the
commit log for a complete list.

### Breaking changes in 4.0

* In `.catch` handlers, `err.response` is now marked as non-enumerable.

### Breaking changes in 3.0

* [Bluebird][bluebird] has been upgraded to version 3.3.1.

### Breaking changes in 2.0

* [Bluebird][bluebird] has been upgraded to version 2.9.24.

[bluebird]: https://github.com/petkaantonov/bluebird
[bluebird-api]: https://github.com/petkaantonov/bluebird/blob/master/API.md#promiseisdynamic-value---boolean
[changelog]: CHANGELOG.md
[expect-api]: https://github.com/visionmedia/supertest#api
[peer-dependency]: http://blog.nodejs.org/2013/02/07/peer-dependencies/
[semver]: http://semver.org
[SuperAgent]: http://visionmedia.github.io/superagent/
[SuperTest]: https://github.com/visionmedia/supertest
[when.js]: https://github.com/cujojs/when


CHANGELOG.md
# Changelog

We follow [semver]: the major version number will be upgraded with any
breaking change. This changelog lists all meaningful changes in each
release; consult the main [README] to determine which changes are
breaking.

## --- / 2016-12-29

  * SuperTest as Promised is deprecated. Thanks, [@jasisk]!

See the README for instructions on upgrading to SuperTest 2.0.

## 4.0.2 / 2016-11-02

  * [[#39]] Bring list of breaking changes in the [README] up-to-date. Thanks, [@karlbecker]!

## 4.0.1 / 2016-10-24

  * expose additional `agent` properties through inheritance chain ([#38])

Thanks, [@jsdevwork]!

## 4.0.0 / 2016-08-11

  * mark `err.response` as non-enumerable to avoid cluttering console output ([#34])

Thanks, [@sylvaingi]!

## 3.2.0 / 2016-07-02

  * attach `response` to error object to ease debugging of failed requests ([#30])

Thanks, [@mkasberg]!

## 3.1.0 / 2016-03-30

  * add `catch()` convenience method directly to test instance ([#23])

Thanks, [@bbatha]!

## 3.0.0 / 2016-02-21

  * update Bluebird to v3.x series

## 2.0.2 / 2015-06-05

  * fix double-resolution of promise upon error ([#11])

Thanks, [@srussellextensis]!

## 2.0.1 / 2015-05-12

  * bump dev dependencies for compatibility with SuperTest's v1.0.0 test
    suite.

## 2.0.0 / 2015-04-06

  * add `toPromise()` method to explicitly convert request to promise
  * support custom promise libraries
  * update Bluebird to v2.x series

## 1.0.0 / 2014-07-01

  * release: v1.0.0
  * support `del` by passing through to upstream's alias
    Closes [#3](https://github.com/WhoopInc/supertest-as-promised/issues/3). Closes [#4](https://github.com/WhoopInc/supertest-as-promised/issues/4).
  * Alias del and delete to supertest.del
  * test: run upstream tests against our wrapped module
  * test: switch to expect interface
  * README: add build badge
  * .travis.yml: add continuous integration configuration
  * refactor `then` method
    `Promise.promisify` is awkward when invoked immediately. Save to
    temporary variable for clarity.
  * CHANGELOG: update for v0.1.1

## 0.1.1 / 2014-05-19

  * release: v0.1.1
  * refactor factory wrapping
    We don't need to inherit from the factory as long as we set
    the receiver correctly when calling its functions. D'oh!
  * CHANGELOG: update for v0.1.0
    v0.0.1 was cut from a rebased tree that was never pushed to GitHub, so
    changelog generation is a bit out-of-sync.
  * release: v0.1.0

## 0.1.0 / 2014-05-19

  * README: fix package name in code example
  * promisify and expose SuperTest agents
    Agents persist cookies across requests. Promisify and expose this
    interface at `exports.agent` to be compatible with SuperTest.
    Fixes [#1](https://github.com/WhoopInc/supertest-as-promised/issues/1).
  * add changelog
  * initial commit

[#39]: https://github.com/WhoopInc/supertest-as-promised/issues/39
[#38]: https://github.com/WhoopInc/supertest-as-promised/issues/38
[#34]: https://github.com/WhoopInc/supertest-as-promised/pull/34
[#30]: https://github.com/WhoopInc/supertest-as-promised/issues/30
[#23]: https://github.com/WhoopInc/supertest-as-promised/pull/23
[#11]: https://github.com/WhoopInc/supertest-as-promised/pull/11

[@bbatha]: https://github.com/bbatha
[@karlbecker]: https://github.com/karlbecker
[@mkasberg]: https://github.com/mkasberg
[@jasisk]: https://github.com/jasisk
[@jsdevwork]: https://github.com/jsdevwork
[@srussellextensis]: https://github.com/srussellextensis
[@sylvaingi]: https://github.com/sylvaingi

[README]: README.md
[semver]: http://semver.org


test/index.js
var expect = require("chai").expect
  , http = require("http")
  , PromiseWhen = require("when").Promise
  , supertest = require("supertest")
  , supertestAsPromised = require("..");

function PromiseMock() {
  var self = this;
  ["then", "catch"].forEach(function (fn) {
    self[fn] = function () {
      this[fn].args.push([].slice.call(arguments));
      return this;
    }
    self[fn].args = [];
  })
}
PromiseMock.resolve = PromiseMock.reject = function () {};

var server = http.createServer(function (req, res) {
  res.end("helo");
});

describe("supertestAsPromised", function () {
  var request = supertestAsPromised(server);

  describe("Test instances", function () {
    describe("#toPromise", function () {
      it("should return a bluebird promise by default", function () {
        expect(request.get("/home").toPromise()).to.be.an.instanceOf(Promise);
      });

      it("should return a when promise if configured with when", function () {
        var request = supertestAsPromised(PromiseWhen)(server);
        expect(request.get("/home").toPromise()).to.be.an.instanceOf(PromiseWhen);
      });

      it("should still return a bluebird promise by default", function () {
        expect(request.get("/home").toPromise()).to.be.an.instanceOf(Promise);
      });
    });

    it("should fulfill if all assertions pass", function () {
      return expect(request.get("/home").expect(200)).to.eventually.be.fulfilled;
    });

    it("should fulfill with the response", function () {
      return request.get("/home").then(function (res) {
        expect(res.text).to.equal("helo");
      });
    });

    it("should reject if an assertion fails", function () {
      return expect(request.get("/home").expect(500)).to.eventually.be.rejected;
    });

    it("should provide response on error object", function () {
      return request.get("/home").expect(500)
        .catch(function (err) {
          return expect(err.response.text).to.equal("helo");
        });
    });

    it("should fulfill with Promise.all when all successful", function () {
      return expect(Promise.all([
        request.get("/home1").expect(200),
        request.get("/home2").expect(200),
      ])).to.eventually.be.fulfilled;
    });

    it("should reject with Promise.all when not successful", function () {
      return expect(Promise.all([
        request.get("/home1").expect(200),
        request.get("/home2").expect(500),
      ])).to.eventually.be.rejected;
    })

    it("should call then with two arguments", function () {
      var mock = supertestAsPromised(PromiseMock)(server).get("/home").then(1, 2, 3);
      expect(mock.then.args).to.deep.equal([[1, 2]]);
    });

    it("should call catch with arbitrary arguments", function () {
      var mock = supertestAsPromised(PromiseMock)(server).get("/home").catch(1, 2, 3);
      expect(mock.catch.args).to.deep.equal([[1, 2, 3]]);
    });
  });

  describe("TestAgent instances", function () {
    var agent = supertestAsPromised.agent(server);

    describe("#toPromise", function () {
      it("should return a promise", function () {
        expect(agent.get("/home").toPromise()).to.be.an.instanceOf(Promise);
      });
    });

    it("should fulfill if all assertions pass", function () {
      return expect(agent.get("/home").expect(200)).to.eventually.be.fulfilled;
    });

    it("should fulfill with the response", function () {
      return agent.get("/home").then(function (res) {
        expect(res.text).to.equal("helo");
      });
    });

    it("should reject if an assertion fails", function () {
      return expect(agent.get("/home").expect(500)).to.eventually.be.rejected;
    });

    it("should be an instance of TestAgent", function () {
      expect(agent).to.be.an.instanceof(supertest.agent);
    });
  });
});

describe("supertest", function () {
  describe("Test instances", function () {
    var request = supertest(server);

    it("should not have toPromise", function () {
      request.get("/home").should.not.have.property("toPromise");
    });
  });

  describe("TestAgent instances", function () {
    var agent = supertest.agent(server);

    it("should not have toPromise", function () {
      agent.get("/home").should.not.have.property("toPromise");
    });
  });
});


test/setup.js
var chai = require("chai")
  , chaiAsPromised = require("chai-as-promised");

chai.should();
chai.use(chaiAsPromised);


index.js
var methods = require("methods")
  , supertest = require("supertest")
  , util = require("util");

// Support SuperTest's historical `del` alias for `delete`
methods = methods.concat("del");

// Generate a SuperTest as Promised module that returns promise
// instances using the provided `Promise` constructor.
function makeModule(Promise) {
  var out;

  function toPromise() {
    var self = this;
    return new Promise(function (resolve, reject) {
      self.end(function (err, res) {
        if (err) {
          // Attach response to error object so that details about the failure
          // (e.g., the response body) can be accessed by a `.catch` handler
          // (#30).  Use `Object.defineProperty` so that `.response` is
          // non-enumerable; otherwise, the error message is lost in the sea of
          // the response object (#34).
          Object.defineProperty(err, 'response', { value: res });
          reject(err);
          return;
        }
        resolve(res);
      });
    });
  }

  function then(onFulfilled, onRejected) {
    return this.toPromise().then(onFulfilled, onRejected);
  }

  function _catch(onRejected) {
    var promise = this.toPromise();
    return promise.catch.apply(promise, arguments);
  }

  // Creates a new object that wraps `factory`, where each HTTP method
  // (`get`, `post`, etc.) is overriden to inject a `then` method into
  // the returned `Test` instance.
  function wrap(factory, out) {
    methods.forEach(function (method) {
      out[method] = function () {
        var test = factory[method].apply(this, arguments);
        test.toPromise = toPromise;
        test.then = then;
        test.catch = _catch;
        return test;
      };
    });

    return out;
  }

  out = function () {
    var request = supertest.apply(null, arguments);
    return wrap(request, {});
  }

  out.agent = function () {
    var self = this;
    if (!(this instanceof out.agent)) {
      self = Object.create(out.agent.prototype);
    }
    supertest.agent.apply(self, arguments);
    return self;
  }
  util.inherits(out.agent, supertest.agent);
  wrap(supertest.agent.prototype, out.agent.prototype);

  return out;
}

// For backwards compatibility, we allow SuperTest as Promised to be
// used without an explicit `Promise` constructor. Pass these requests
// through to a default module that uses Bluebird promises.

var defaultModule = makeModule(Promise);

module.exports = function (maybePromise) {
  if (typeof maybePromise.resolve === 'function' &&
      typeof maybePromise.reject === 'function') {
    return makeModule(maybePromise);
  }

  return defaultModule.apply(null, arguments);
}

module.exports.agent = defaultModule.agent;
